### PR TITLE
heffte: new oneAPI variant

### PR DESCRIPTION
**Pull request by mabraham**

### CODE REVIEW

File: var/spack/repos/builtin/packages/heffte/package.py [hunk 1]
```diff
@@ -38,6 +38,7 @@ class Heffte(CMakePackage, CudaPackage, ROCmPackage):
 
     variant("shared", default=True, description="Builds with shared libraries")
     variant("fftw", default=False, description="Builds with support for FFTW backend")
+    variant("oneapi", default=False, description="Builds with support for oneAPI backend")
```
Comment (mkstoyanov) [inline var/spack/repos/builtin/packages/heffte/package.py:41]: @mabraham Are you using heFFTe + OneAPI + spack?

The reason why I was holding back on this one is that I'm not sure what the name should be. When CUDA was the only GPU platform, people were using variants, such as `+gpu` or `+cuda`. Later spack standardized `+cuda` but all the `+gpu` packages had issues with backwards compatibility. Later, we had the new platform from AMD, but then people used variants `+hip` and `+rocm`, and only later spack standardized `+rocm`.

How should we call this option now? Spack has no standard for OneAPI, as far as I know. Several other packages are already using the `+sycl` variant, so should we use `+sycl` or `+oneapi` or `+intel_gpu` or `+dpc++` or `+dpcpp`.

In short, I was holding on this feature until spack decides what the equivalent of `+cuda` and `+rocm` should be for the Intel platform.

https://spack.readthedocs.io/en/latest/build_systems/cudapackage.html

https://spack.readthedocs.io/en/latest/build_systems/rocmpackage.html

If you're actually using heFFTe + OneAPI + spack then we can add this and deal with the consequences later. I would still change the name of the variant to `+sycl` but we can work with it for now. If not, I think it is best to take our time.

Comment (mabraham) [reply var/spack/repos/builtin/packages/heffte/package.py:41]: Yes, I'm Intel's application engineer supporting GROMACS, and its strong-scaling support for e.g. Aurora and friends uses HeFFTe on Intel MPI and MKL. I have such a GROMACS build working with spack and after I upstream this and some other pieces, I will use that combination in my daily work and recommend it to my peers.

HeFFTe already has other CUDA and ROCm support, so it's unlikely to want SYCL support for any compiler other than that from oneAPI, so I think `+oneapi` is accurate as far as HeFFTe goes. I think that fundamentally the mixin aproach suits CUDA and ROCm for use with Spack, but I don't see the same approach being viable for oneAPI and think that a dependency on the oneAPI compiler is the most usable path. They do all kind of depend on a/another compiler though... I think we need to open an issue with a larger audience to get traction on that question, though!

GROMACS currently uses `+sycl` in its recipe, but only hipSYCL (now called AdaptiveCpp) `provides(sycl)`. I intend to change the GROMACS recipe so that `gromacs +sycl` can be satisfied either by a package that `provides(sycl)` or `%oneapi` (which seems to work in my hands), but first I have some simpler changes to the GROMACS recipe to do. I mention this because if HeFFTe has plans to support another SYCL compiler also, or you really want to use `+sycl`, then you might want the same approach.

Comment (mabraham) [reply var/spack/repos/builtin/packages/heffte/package.py:41]: An argument against `+oneapi` is that in theory you could implement HeFFTe with GPU-aware MPI and OpenMP-offloading. At that time, one would want `%oneapi +openmp` to describe that HeFFTe implementation and `%oneapi +sycl` to describe the existing one based on SYCL.

As dpcpp is deprecated, don't use that.

Comment (mkstoyanov) [reply var/spack/repos/builtin/packages/heffte/package.py:41]: OK, we will enable this one way or another.

When it comes to SYCL and heFFTe, the compiler is not the issue. The problem is that the compiler has to be paired with the proper FFT library (e.g., cuFFT/rocFFT/oneMKL), and currently we only support SYCL + oneMKL (CUDA with cuFFT, and so on).

My main concern about the name is not related to the relationship between SYCL, DPC++ and oneAPI, or the availability of the compilers. My concern is with spack, sooner or later they will standardize the option just as they did with `+cuda` and `+rocm`. If I have to guess, I would say that they will use `+sycl` and that's the only reason why I want to change the name.

Comment (mkstoyanov) [reply var/spack/repos/builtin/packages/heffte/package.py:41]: Anyhow, this is no reason to hold back the PR

**Review by mkstoyanov — APPROVED** [review]

File: var/spack/repos/builtin/packages/heffte/package.py (hunk 1 above)
Comment (mabraham) [reply var/spack/repos/builtin/packages/heffte/package.py:41]: OK. I agree they're more likely to standardize to `+sycl` than anything else, and I only picked `+oneapi` in the first place to match the name of HeFFTe's CMake option. so I'll upload a new PR tomorrow with `+sycl` and you can accept whichever seems better then.

Comment (mabraham) [reply var/spack/repos/builtin/packages/heffte/package.py:41]: I suggest we adopt #41132 instead

Comment (mkstoyanov) [reply var/spack/repos/builtin/packages/heffte/package.py:41]: I also vote in favor of #41132 

You should probably close this one.

**Comment by eugeneswalker** [comment]
Closing in favor of now-merged https://github.com/spack/spack/pull/41132
